File: Search2DArray.java
// Time Complexity : O(log m + log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
// 1. Binary search on the rows to locate the row that might contain the target.
// 2. Binary search on that row to locate the target within the columns.


public class Search2DArray {
    public boolean searchMatrix(int[][] matrix, int target) {
        int length = matrix.length;
        int innerLength = matrix[0].length;

        int low = 0; int high = length - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (matrix[mid][0] <= target && matrix[mid][innerLength - 1] >= target) {
                low = 0;
                high = innerLength - 1;
                while (low <= high) {
                    int innerMid = low + (high - low) / 2;
                    if (matrix[mid][innerMid] == target) return true;
                    if (matrix[mid][innerMid] > target) {
                        high = innerMid - 1;
                    } else {
                        low = innerMid + 1;
                    }
                }
                return false;
            }

            if (matrix[mid][0] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return false;
    }
}

File: search2Darray.py
# Time Complexity : O(log(m*n))
# Space Complexity : O(log(m*n))
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : difficulty in assigning low and high pointers in the 2D array like took some time to figure that out


# Your code here along with comments explaining your approach in three sentences only

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix) # calculating the number of rows
        cols = len(matrix[0]) # calculating the number of columns

        low = 0
        high = rows * cols - 1

        # performing binary search
        while low <= high:
            mid = (low + high) // 2
            r = mid//cols # calculating rows of the 2D matrix
            c = mid%cols # calculating the columns of the 2D matrix

            if matrix[r][c] == target:
                return True

            elif matrix[r][c] < target:
                low = mid + 1

            else:
                high = mid - 1

        return False
